### PR TITLE
Improve mount performance by bulk attachment

### PR DESCRIPTION
**Pull request by lttb**
We have some overhead on `detach` and `attach().link()` each mount, so when we have a lot of components that we need to mount at the same time (like https://github.com/A-gambit/CSS-IN-JS-Benchmarks), there some issues.

This fix make significant performance improvement by bulk updating (more then **20x** in example above).

*TODO*:
- [x] try [asap](https://github.com/kriskowal/asap), it may be faster
> `asap` and `raf` have similar performance in bench above, but `asap` was little bit faster and had more stable results
> UPDATE: but it seems that better to use `raf` for loops
- [ ] check css floating with styled elements created on render
- [ ] add tests to https://github.com/A-gambit/CSS-IN-JS-Benchmarks

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -42,7 +42,8 @@
   "dependencies": {
     "is-react-prop": "^0.0.3",
     "jss": "^7.1.0",
-    "jss-preset-default": "^2.0.0"
+    "jss-preset-default": "^2.0.0",
+    "raf": "^3.3.0"
```
Comment (kof) [inline package.json]: I think raf polyfill should be responsibility of the user, like any other new features.

Comment (kof) [inline package.json]: So I would just use requestAnimationFrame, its well implemented right now anyways. All required polyfills should be in the readme

File: src/index.js [hunk 2]
```diff
@@ -19,6 +20,15 @@ const createStyled = (jss?: Function = jssDefault) => (baseStyles: Object = {})
   let dynamicSheet
   let counter = 0
 
+  let dynamicCounter = 0
+  raf(function update() {
+    if (dynamicCounter) {
+      dynamicCounter = 0
+      dynamicSheet.attach().link()
```
Comment (kof) [inline src/index.js]: better call link() first

File: package.json (hunk 1 above)
Comment (lttb) [inline package.json]: We can't use just `window.requestAnimationFrame`, because we run this code not only in the browser (tests, ssr) - so we need polyfill

But we can make it customizable

**Comment by kof** [comment]
Tests should implement a mock if they need it.

**Comment by lttb** [comment]
> Tests should implement a mock if they need it.

you are right, but the main point is ssr :)

**Comment by kof** [comment]
mock it in ssr or don't use it when on the server

File: src/createStyled.js [hunk 3]
```diff
@@ -9,11 +10,25 @@ import type {
...
 const createStyled = (jss: Function) => (
   baseStyles: BaseStylesType = {}
 ): StyledType => {
   let staticSheet
   let dynamicSheet
+  let dynamicSheetId
+
+  const addRule = (name: string, style: ComponentStyleType, data: Object) => {
+    if (data) {
+      dynamicSheet.detach().addRule(name, style)
```
Comment (kof) [inline src/createStyled.js]: needs a comment, its because of a bug, right?

File: src/utils/sheetsObserver.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      dynamicSheets[Number(sheetId)].attach().link())
+  }
+}
+
+const listen = () => {
+  observe()
+  raf(listen)
+}
+
+export default {
```
Comment (kof) [inline src/utils/sheetsObserver.js]: its not really an observer ... also it might be confused with a real Observer .

**Comment by lttb** [comment]
The main problem of this fix is async styles attaching, that may cause FUC (flash of unstyled content). 
So we can return to this later.

Problem with mount performance solved in #12
